### PR TITLE
Fix search result pagination

### LINKED ISSUE #4027: Chef::Search::Query#search might return duplicate results or loop forever
PR #3135 introduced a bug where search() might list results more than once or never complete.

This issue is here:

https://github.com/chef/chef/blob/master/lib/chef/search/query.rb#L92

Incrementing by the count of the rows you got back seems correct, but the problem is that the start and rows parameters are passed directly to solr on the backend.  Results from solr, however, might included deleted nodes that are filtered from the results.  Thus, you might ask for the first page of 10 results, get back 8 nodes because 2 of the 10 results returned by solr were filtered.  However, you should still start your next page at 10 because the "pages" that matter are the results coming back from solr.

NB: The author of the PR original change would have been fine. Unfortunately, I gave a braindead recommendation that led to this bug.

Fix incoming.

### PR DESCRIPTION
**Pull request by stevendanna**
The start and rows parameter that are passed as part of the search
request are passed directly to Solr on the backend. Results from Solr
may contain deleted nodes no longer in the erchef database.  Erchef will
filter such nodes from the results. Thus, a user may receive fewer rows
than they asked for. Incrementing 'start' only by the number of rows
received will then result in the next Solr response overlapping the
first, which can lead to duplicate results. In the case of a Solr
results page that was completely filtered, it would lead to an infinite
loop.

This commit changes the code to always increment by the requested page
size (args[:rows]) when it is non-nil. Incrementing by the length of the
response set is still wrong in the case when the args[:rows] is nil, but
the server doesn't give us anything else to increment by.

Fixes #4027

### CODE REVIEW

**Comment by stevendanna** [comment]
cc @chef/client-core 

This is still problematic in the case where the user doesn't pass a rows attribute.  Perhaps we want to deprecate searches without rows specified?

**Comment by danielsdeleo** [comment]
:+1:

File: lib/chef/search/query.rb [hunk 1]
```diff
@@ -88,8 +88,9 @@ def search(type, query='*:*', *args, &block)
 
         if block
           response["rows"].each { |row| block.call(row) if row }
-          unless (response["start"] + response["rows"].length) >= response["total"]
-            args_h[:start] = response["start"] + response["rows"].length
+          next_start = response["start"] + (args_h[:rows] || response["rows"].length)
+          unless next_start >= response["total"]
```
Comment (lamont-granquist) [inline lib/chef/search/query.rb]: might want to have a comment here to explain these fields correctly.

**Comment by lamont-granquist** [comment]
:+1:

File: lib/chef/search/query.rb (hunk 1 above)
Comment (danielsdeleo) [inline lib/chef/search/query.rb]: good point, this is a good place to make a note that the results may not be the same size as the amount of rows we asked for.

**Comment by stevendanna** [comment]
Added a lengthy comment.

**Comment by jaym** [comment]
excellent comment :+1:

**Comment by danielsdeleo** [comment]
Merge whenever.

**Pull request merged**